File: homework/tai_nguyen/lesson7.2/1730_shortest_path_to_get_food.ts
const dr = [-1, 1, 0, 0];
const dc = [0, 0, -1, 1];

function isInsideGrid(grid: string[][], i: number, j: number) {
  const m = grid.length;
  const n = grid[0].length;

  return i >= 0 && i < m && j >= 0 && j < n;
}

function getFood(grid: string[][]): number {
  function bfs(row: number, col: number) {
    const queue: [number, number, number][] = [[row, col, 0]];

    while (queue.length) {
      const [i, j, level] = queue.shift()!;

      for (let k = 0; k < 4; k++) {
        const x = i + dr[k];
        const y = j + dc[k];

        if (isInsideGrid(grid, x, y) && grid[x][y] !== 'X') {
          if (grid[x][y] === '#') {
            return level + 1;
          }

          grid[x][y] = 'X'; // marked as visited
          queue.push([x, y, level + 1]);
        }
      }
    }

    return -1;
  }

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++) {
      if (grid[i][j] === '*') {
        grid[i][j] = 'X'; // marked as visited
        return bfs(i, j);
      }
    }
  }

  return -1;
}

File: homework/thinhbka/lesson5-graph/1730. Shortest Path to Get Food.py
DIR = [(0,-1),(0,1),(1,0),(-1,0)]
class Solution:
    def getFood(self, grid: List[List[str]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        visited = [[False for _ in range(cols)] for _ in range(rows)]
        step = [[0 for _ in range(cols)] for _ in range(rows)]
        has_found_star = False
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == "*" and has_found_star is False:
                    has_found_star = True
                    visited[i][j] = True
                    frontier = [(i,j)]
                    while frontier:
                        next = []
                        for (row,col) in frontier:
                            for (m,n) in DIR:
                                n_r = row+m
                                n_c = col+n
                                if 0<=n_r<rows and 0<=n_c<cols and grid[n_r][n_c] != "X" and visited[n_r][n_c] is False:
                                    step[n_r][n_c] = step[row][col]+1
                                    visited[n_r][n_c] = True
                                    if grid[n_r][n_c] =="#":
                                        return step[n_r][n_c]
                                    next.append((n_r,n_c))
                        frontier = next
        return -1